apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,309 @@
+    setSyncedCollectionMtime: setSyncedCollectionMtime,
+    getSyncedCollectionMtime: getSyncedCollectionMtime,
+    updatePlaces: updatePlaces,
+    deletePlaceByFxSyncId: deletePlaceByFxSyncId
+  };
+})();
+
+DataAdapters.history = {
+/**
+    KintoCollection.list() provides a list contains all records sorted by
+  "last_modified" property which indicates the data-modified time in
+  descending order. The "last_modified" value can be recongized as which
+  records have already synchronized before. It is used to stop early and avoid
+  the redundant operations which had handled.
+
+    When the new adding/updating records are coming from History Collection in
+  the format [2], it should be converted to the format [1] then stored into PDS.
+  Besides, the ID from HC records should be written into PDS records as well
+  for deleting record usage.
+
+    When the new deleting records are coming, the ID can help us to find which
+  record has to be removed.
+
+  1. Records stored in Places DataStore (PDS): {
+    "url": "http://mozilla.org/", // KEY in PDS
+    "title": "Mozilla",
+    "icons": {
+      "http://mozilla.org/favicon.ico": {
+        "sizes": []
+      }
+    },
+    "frecency": 1,
+    "visits": [
+      // NOTICE: date/time without ms
+      1442247252490, 1442247250001
+    ],
+    "screenshot": {},
+    "visited": 1442247252490
+  }
+
+  2. Add/Update Records from History Collection (HC): {
+    "id": "zMgfGkRinh92",
+    "sortindex": 2000,
+    "last_modified": 1442247272150,
+    "payload": {
+      "id": "zMgfGkRinh92",
+      "histUri": "http://mozilla.org/",
+      "title": "Mozilla",
+      "visits": [
+         // NOTICE: date/time with ms
+        { "date": 1442247252490018, "type": 2 },
+        { "date": 1442247250001234, "type": 2 }
+      ]
+    },
+    "_status": "synced"
+  }
+
+  3. Delete Records from History Collection (HC): {
+    "id": "_Avscjx5srFy",
+    "sortindex": 100,
+    "last_modified": 1441985077970,
+    "payload": {
+      "id": "_Avscjx5srFy",
+      "deleted": true
+    },
+    "_status": "synced"
+  }
+**/
+  _fullSync(kintoCollection, lastModifiedTime) {
+    function updateHistoryCollection(list) {
+      var historyRecords = list.data;
+      var places = [];
+      var i;
+      for (i = 0; i < historyRecords.length; i++) {
+        if (historyRecords[i].last_modified <= lastModifiedTime) {
+          break;
+        }
+      }
+      var partialRecords = historyRecords.slice(0, i);
+      if (partialRecords.length === 0) {
+        return Promise.resolve(false);
+      }
+      partialRecords.forEach(record => {
+        var payload = record.payload;
+        if (payload.deleted) {
+          places.push({
+            deleted: true,
+            fxsyncId: payload.id
+          });
+          return;
+        }
+        if (!payload.histUri || !payload.visits || !payload.visits.length) {
+          console.warn('Incorrect payload? ', JSON.stringify(payload));
```

Comment:
nit: I think you don't need to JSON.stringify anymore, the console already takes care of that for you.